In [1]:
import pandas as pd
import numpy as np

/home/taina/.local/lib/python3.7/site-packages/pandas/compat/_optional.py:138: UserWarning: Pandas requires version '2.7.0' or newer of 'numexpr' (version '2.6.9' currently installed).
  warnings.warn(msg, UserWarning)


In [3]:
crawl = pd.read_csv('~/Downloads/projeto_malwares_android-main/Datasets/AndroCrawl/AndroCrawl.csv')

/home/taina/.local/lib/python3.7/site-packages/IPython/core/interactiveshell.py:3552: DtypeWarning: Columns (27,28,29,30,31,32,33,35,36,37,38,39,40,41,42,43,44,45,46,47,48,49,50,51,52,53,54,55,56,57,58,59,60,61,62,63,64,65,66,67,68,69,70,71,72,73,74,75,76,77,78,79,80,81,82,83,84,85,86,87,88,89,90,91,92,93,94,95,98,99,101,103,106,109,110,111,112,117,118,119,123,124,125,126,127,128,130,131,132,133,134,135,136,137,138,139,140,141,142,143,144,145,146,147,148,149,150,151,152,153,154,155,188,189,190,191,192,193,194,195,197,198,199,200,201,202,203,204,206,207,208,209,210,211,212,213,215,216,217,218,220) have mixed types.Specify dtype option on import or set low_memory=False.
  exec(code_obj, self.user_global_ns, self.user_ns)


In [4]:
crawl

'Apk Name'  \
0       51113d15a58841b8c81bc7fc8fd7b45f4ca19a29029b86...   
1       3f944d58fbb88b8c64c82c3cd60b12d851176df8aca0d2...   
2       1370ea8b3e2d895b83b5a7771348546eb397624cdc52aa...   
3       e0ac1ffb30426d6f650fa7b3244277f273220fbe230ef6...   
4       0ede82b2a99ed040e65ce1816f054921187d184320301f...   
...                                                   ...   
162978  559f0610501a1bd0f3e7265494cec8537bb0d8c2da47bb...   
162979  a4f11ba03162b949aa29221d5bb2e4909bb116e8cc41a9...   
162980  1f3d156bf6ac206e4cf79db74065865eaf414f181b3bcd...   
162981  cc4b6e06f05a34779803e5f0d25b226a0a9cd10ba5453c...   
162982  eda85939db98c9f5283ce9304f40ea1f841f25a8586b13...   

        'Dangerous Permission: ACCESS_SUPERUSER'  \
0                                          False   
1                                          False   
2                                          False   
3                                          False   
4                                          False   
...                                          ...   
162978                                     False   
162979                                     False   
162980                                     False   
162981                                     False   
162982                                     False   

        'Dangerous Permission: BLUETOOTH_PRIVILEGED'  \
0                                              False   
1                                              False   
2                                              False   
3                                              False   
4                                              False   
...                                              ...   
162978                                         False   
162979                                         False   
162980                                         False   
162981                                         False   
162982                                         False   

        'Dangerous Permission: BRICK'  \
0                               False   
1                               False   
2                               False   
3                               False   
4                               False   
...                               ...   
162978                          False   
162979                          False   
162980                          False   
162981                          False   
162982                          False   

        'Dangerous Permission: CHANGE_COMPONENT_ENABLED_STATE'  \
0                                                   False        
1                                                   False        
2                                                   False        
3                                                   False        
4                                                   False        
...                                                   ...        
162978                                              False        
162979                                              False        
162980                                              False        
162981                                              False        
162982                                              False        

        'Dangerous Permission: CLEAR_APP_USER_DATA'  \
0                                             False   
1                                             False   
2                                             False   
3                                             False   
4                                             False   
...                                             ...   
162978                                        False   
162979                                        False   
162980                                        False   
162981                                        False   
162982                                        False   

        'Dangerous Permission: DELETE_CACHE_FILES'  \
0              

In [5]:
list(crawl.columns)

["'Apk Name'",
 "'Dangerous Permission: ACCESS_SUPERUSER'",
 "'Dangerous Permission: BLUETOOTH_PRIVILEGED'",
 "'Dangerous Permission: BRICK'",
 "'Dangerous Permission: CHANGE_COMPONENT_ENABLED_STATE'",
 "'Dangerous Permission: CLEAR_APP_USER_DATA'",
 "'Dangerous Permission: DELETE_CACHE_FILES'",
 "'Dangerous Permission: DELETE_PACKAGES'",
 "'Dangerous Permission: DISABLE_KEYGUARD'",
 "'Dangerous Permission: FACTORY_TEST'",
 "'Dangerous Permission: INSTALL_PACKAGES'",
 "'Dangerous Permission: INJECT_EVENTS'",
 "'Dangerous Permission: INTERNAL_SYSTEM_WINDOW'",
 "'Dangerous Permission: KILL_BACKGROUND_PROCESSES'",
 "'Dangerous Permission: MASTER_CLEAR'",
 "'Dangerous Permission: MODIFY_PHONE_STATE'",
 "'Dangerous Permission: MOUNT_FORMAT_FILESYSTEM'",
 "'Dangerous Permission: MOUNT_UNMOUNT_FILESYSTEM'",
 "'Dangerous Permission: PROCESS_OUTGOING_CALLS'",
 "'Dangerous Permission: READ_LOGS'",
 "'Dangerous Permission: REBOOT'",
 "'Dangerous Permission: RECEIVE_BOOT_COMPLETED'",
 "'Dangerous 

In [6]:
col = list(crawl.columns)

### Lista de permissões:
As permissões de um app ajudam a apoiar a privacidade do usuário protegendo dados restritos e ações restritas

In [7]:
permissions = []
for i in col:
    if "Permission" in i:
            permissions.append(i)

permissions

["'Dangerous Permission: ACCESS_SUPERUSER'",
 "'Dangerous Permission: BLUETOOTH_PRIVILEGED'",
 "'Dangerous Permission: BRICK'",
 "'Dangerous Permission: CHANGE_COMPONENT_ENABLED_STATE'",
 "'Dangerous Permission: CLEAR_APP_USER_DATA'",
 "'Dangerous Permission: DELETE_CACHE_FILES'",
 "'Dangerous Permission: DELETE_PACKAGES'",
 "'Dangerous Permission: DISABLE_KEYGUARD'",
 "'Dangerous Permission: FACTORY_TEST'",
 "'Dangerous Permission: INSTALL_PACKAGES'",
 "'Dangerous Permission: INJECT_EVENTS'",
 "'Dangerous Permission: INTERNAL_SYSTEM_WINDOW'",
 "'Dangerous Permission: KILL_BACKGROUND_PROCESSES'",
 "'Dangerous Permission: MASTER_CLEAR'",
 "'Dangerous Permission: MODIFY_PHONE_STATE'",
 "'Dangerous Permission: MOUNT_FORMAT_FILESYSTEM'",
 "'Dangerous Permission: MOUNT_UNMOUNT_FILESYSTEM'",
 "'Dangerous Permission: PROCESS_OUTGOING_CALLS'",
 "'Dangerous Permission: READ_LOGS'",
 "'Dangerous Permission: REBOOT'",
 "'Dangerous Permission: RECEIVE_BOOT_COMPLETED'",
 "'Dangerous Permission: STA

In [8]:
dangerous = []
c2m = []
harmless = []

for i in permissions:
    if "Dangerous" in i:
      dangerous.append(i)

    if "Harmless" in i:
      harmless.append(i)
    
    if "C2M" in i:
      c2m.append(i)

print(len(dangerous), len(c2m), len(harmless))

26 3 32


**Lista de permissões: 61 permissões no total**


*   26 Dangerous Permission
*   3 C2M Permission
*   32 Harmless Permission

### Lista de Chamadas de API:
API Calls: As requisições de API são formas de fazer um software conversar com outro e permitir a troca de informações entre eles.

In [10]:
APICalls = []
for i in col:
  if "Api" in i:
    APICalls.append(i)
    
APICalls

["'Api Call: Landroid/content/Context;->startService'",
 "'Api Call: Landroid/telephony/TelephonyManager;->getSubscriberId'",
 "'Api Call: Landroid/telephony/TelephonyManager;->getDeviceId'",
 "'Api Call: Landroid/telephony/TelephonyManager;->getLine1Number'",
 "'Api Call: Landroid/telephony/TelephonyManager;->getSimSerialNumber'",
 "'Api Call: Landroid/telephony/TelephonyManager;->getSimOperatorName'",
 "'Api Call: Landroid/telephony/TelephonyManager;->getCellLocation'",
 "'Api Call: Landroid/telephony/cdma/CdmaCellLocation;->getSystemId'",
 "'Api Call: Landroid/telephony/SmsManager;->sendTextMessage'",
 "'Api Call: Landroid/content/Intent;->setDataAndType'",
 "'Api Call: Landroid/content/Intent;->setType'",
 "'Api Call: Landroid/app/ActivityManager;->getRunningServices'",
 "'Api Call: Landroid/app/ActivityManager;->getMemoryInfo'",
 "'Api Call: Landroid/app/ActivityManager;->restartPackage'",
 "'Api Call: Landroid/content/pm/PackageManager;->getInstalledPackages'",
 "'Api Call: Ljava

In [11]:
no_APICalls = []
for i in APICalls:
  if "Notification" in i:
    no_APICalls.append(i)

for i in no_APICalls:
  APICalls.remove(i)


APICalls

["'Api Call: Landroid/content/Context;->startService'",
 "'Api Call: Landroid/telephony/TelephonyManager;->getSubscriberId'",
 "'Api Call: Landroid/telephony/TelephonyManager;->getDeviceId'",
 "'Api Call: Landroid/telephony/TelephonyManager;->getLine1Number'",
 "'Api Call: Landroid/telephony/TelephonyManager;->getSimSerialNumber'",
 "'Api Call: Landroid/telephony/TelephonyManager;->getSimOperatorName'",
 "'Api Call: Landroid/telephony/TelephonyManager;->getCellLocation'",
 "'Api Call: Landroid/telephony/cdma/CdmaCellLocation;->getSystemId'",
 "'Api Call: Landroid/telephony/SmsManager;->sendTextMessage'",
 "'Api Call: Landroid/content/Intent;->setDataAndType'",
 "'Api Call: Landroid/content/Intent;->setType'",
 "'Api Call: Landroid/app/ActivityManager;->getRunningServices'",
 "'Api Call: Landroid/app/ActivityManager;->getMemoryInfo'",
 "'Api Call: Landroid/app/ActivityManager;->restartPackage'",
 "'Api Call: Landroid/content/pm/PackageManager;->getInstalledPackages'",
 "'Api Call: Ljava

In [12]:
len(APICalls)

27

**Lista de API Calls: 27 chamadas no total**

### Lista de Intenções:

In [13]:
intents = []
for i in col:
  if "Intent" in i:
    intents.append(i)

intents

["'Api Call: Landroid/content/Intent;->setDataAndType'",
 "'Api Call: Landroid/content/Intent;->setType'",
 "'Api Call: Landroid/content/Intent;->describeContents'",
 "'C2M Intent: c2dm.intent.RECEIVE'",
 "'C2M Intent: c2dm.intent.REGISTRATION'",
 "'C2M Intent: c2dm.intent.REGISTER'",
 "'Suspicious Intent Filter: AIRPLANE_MODE_CHANGED'",
 "'Suspicious Intent Filter: BOOT_COMPLETED'",
 "'Suspicious Intent Filter: CONFIGURATION_CHANGED'",
 "'Suspicious Intent Filter: BATTERY_CHANGED'",
 "'Suspicious Intent Filter: DEVICE_STORAGE_LOW'",
 "'Suspicious Intent Filter: DOCK_EVENT'",
 "'Suspicious Intent Filter: EXTERNAL_APPLICATIONS_AVAILABLE'",
 "'Suspicious Intent Filter: MANAGE_PACKAGE_STORAGE'",
 "'Suspicious Intent Filter: MEDIA_MOUNTED'",
 "'Suspicious Intent Filter: MY_PACKAGE_REPLACED'",
 "'Suspicious Intent Filter: NEW_OUTGOING_CALL'",
 "'Suspicious Intent Filter: PACKAGE_ADDED'",
 "'Suspicious Intent Filter: PACKAGE_CHANGED'",
 "'Suspicious Intent Filter: PACKAGE_DATA_CLEARED'",
 "'

In [14]:
no_intents = []
for i in intents:
  if "Api" in i:
    no_intents.append(i)

for i in no_intents:
  intents.remove(i) 

intents

["'C2M Intent: c2dm.intent.RECEIVE'",
 "'C2M Intent: c2dm.intent.REGISTRATION'",
 "'C2M Intent: c2dm.intent.REGISTER'",
 "'Suspicious Intent Filter: AIRPLANE_MODE_CHANGED'",
 "'Suspicious Intent Filter: BOOT_COMPLETED'",
 "'Suspicious Intent Filter: CONFIGURATION_CHANGED'",
 "'Suspicious Intent Filter: BATTERY_CHANGED'",
 "'Suspicious Intent Filter: DEVICE_STORAGE_LOW'",
 "'Suspicious Intent Filter: DOCK_EVENT'",
 "'Suspicious Intent Filter: EXTERNAL_APPLICATIONS_AVAILABLE'",
 "'Suspicious Intent Filter: MANAGE_PACKAGE_STORAGE'",
 "'Suspicious Intent Filter: MEDIA_MOUNTED'",
 "'Suspicious Intent Filter: MY_PACKAGE_REPLACED'",
 "'Suspicious Intent Filter: NEW_OUTGOING_CALL'",
 "'Suspicious Intent Filter: PACKAGE_ADDED'",
 "'Suspicious Intent Filter: PACKAGE_CHANGED'",
 "'Suspicious Intent Filter: PACKAGE_DATA_CLEARED'",
 "'Suspicious Intent Filter: PACKAGE_FIRST_LAUNCH'",
 "'Suspicious Intent Filter: PACKAGE_INSTALL'",
 "'Suspicious Intent Filter: PACKAGE_REMOVED'",
 "'Suspicious Intent

In [15]:
c2mIntents = []
suspicious = []

for i in intents:
  if "C2M" in i:
    c2mIntents.append(i)
  
  if "Suspicious" in i:
    suspicious.append(i)

print(len(c2mIntents), len(suspicious))

3 24


**Lista de Intenções: 27 intenções ao total**
*   3 C2M Intent
*   24 Suspicious Intent Filter

### Lista de todas as colunas de interesse: 

In [16]:
crawlResultList = permissions + APICalls + intents

In [17]:
crawlResultList

["'Dangerous Permission: ACCESS_SUPERUSER'",
 "'Dangerous Permission: BLUETOOTH_PRIVILEGED'",
 "'Dangerous Permission: BRICK'",
 "'Dangerous Permission: CHANGE_COMPONENT_ENABLED_STATE'",
 "'Dangerous Permission: CLEAR_APP_USER_DATA'",
 "'Dangerous Permission: DELETE_CACHE_FILES'",
 "'Dangerous Permission: DELETE_PACKAGES'",
 "'Dangerous Permission: DISABLE_KEYGUARD'",
 "'Dangerous Permission: FACTORY_TEST'",
 "'Dangerous Permission: INSTALL_PACKAGES'",
 "'Dangerous Permission: INJECT_EVENTS'",
 "'Dangerous Permission: INTERNAL_SYSTEM_WINDOW'",
 "'Dangerous Permission: KILL_BACKGROUND_PROCESSES'",
 "'Dangerous Permission: MASTER_CLEAR'",
 "'Dangerous Permission: MODIFY_PHONE_STATE'",
 "'Dangerous Permission: MOUNT_FORMAT_FILESYSTEM'",
 "'Dangerous Permission: MOUNT_UNMOUNT_FILESYSTEM'",
 "'Dangerous Permission: PROCESS_OUTGOING_CALLS'",
 "'Dangerous Permission: READ_LOGS'",
 "'Dangerous Permission: REBOOT'",
 "'Dangerous Permission: RECEIVE_BOOT_COMPLETED'",
 "'Dangerous Permission: STA

## Dataset Final:

In [18]:
crawl["'Detection Ratio'"].value_counts()

Benign       139670
Malicious     23313
Name: 'Detection Ratio', dtype: int64

In [19]:
crawlResult = crawl[crawlResultList]

In [20]:
crawlResult["'Detection Ratio'"] = crawl["'Detection Ratio'"]

/home/taina/.local/lib/python3.7/site-packages/ipykernel_launcher.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  """Entry point for launching an IPython kernel.


In [21]:
crawlResult["'Detection Ratio'"] = crawlResult["'Detection Ratio'"].replace("Malicious",int(1))
crawlResult["'Detection Ratio'"] = crawlResult["'Detection Ratio'"].replace("Benign",int(0))

/home/taina/.local/lib/python3.7/site-packages/ipykernel_launcher.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  """Entry point for launching an IPython kernel.
/home/taina/.local/lib/python3.7/site-packages/ipykernel_launcher.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  


In [22]:
crawlResult

'Dangerous Permission: ACCESS_SUPERUSER'  \
0                                          False   
1                                          False   
2                                          False   
3                                          False   
4                                          False   
...                                          ...   
162978                                     False   
162979                                     False   
162980                                     False   
162981                                     False   
162982                                     False   

        'Dangerous Permission: BLUETOOTH_PRIVILEGED'  \
0                                              False   
1                                              False   
2                                              False   
3                                              False   
4                                              False   
...                                              ...   
162978                                         False   
162979                                         False   
162980                                         False   
162981                                         False   
162982                                         False   

        'Dangerous Permission: BRICK'  \
0                               False   
1                               False   
2                               False   
3                               False   
4                               False   
...                               ...   
162978                          False   
162979                          False   
162980                          False   
162981                          False   
162982                          False   

        'Dangerous Permission: CHANGE_COMPONENT_ENABLED_STATE'  \
0                                                   False        
1                                                   False        
2                                                   False        
3                                                   False        
4                                                   False        
...                                                   ...        
162978                                              False        
162979                                              False        
162980                                              False        
162981                                              False        
162982                                              False        

        'Dangerous Permission: CLEAR_APP_USER_DATA'  \
0                                             False   
1                                             False   
2                                             False   
3                                             False   
4                                             False   
...                                             ...   
162978                                        False   
162979                                        False   
162980                                        False   
162981                                        False   
162982                                        False   

        'Dangerous Permission: DELETE_CACHE_FILES'  \
0                                            False   
1                                            False   
2                                            False   
3                                            False   
4                                            False   
...                                            ...   
162978                                       False   
162979                                       False   
162980                                       False   
162981                                       False   
162982                                       False   

        'Dangerous Permission: DELETE_PACKAGES'  \
0                                         False   
1           

In [23]:
crawlResult["'Detection Ratio'"].value_counts()

0    139670
1     23313
Name: 'Detection Ratio', dtype: int64

In [24]:
crawlResult = crawlResult.mask(crawlResult == False, 0)
crawlResult = crawlResult.mask(crawlResult == 'false', 0)

In [25]:
crawlResult

'Dangerous Permission: ACCESS_SUPERUSER'  \
0                                             0   
1                                             0   
2                                             0   
3                                             0   
4                                             0   
...                                         ...   
162978                                        0   
162979                                        0   
162980                                        0   
162981                                        0   
162982                                        0   

        'Dangerous Permission: BLUETOOTH_PRIVILEGED'  \
0                                                  0   
1                                                  0   
2                                                  0   
3                                                  0   
4                                                  0   
...                                              ...   
162978                                             0   
162979                                             0   
162980                                             0   
162981                                             0   
162982                                             0   

        'Dangerous Permission: BRICK'  \
0                                   0   
1                                   0   
2                                   0   
3                                   0   
4                                   0   
...                               ...   
162978                              0   
162979                              0   
162980                              0   
162981                              0   
162982                              0   

       'Dangerous Permission: CHANGE_COMPONENT_ENABLED_STATE'  \
0                                                       0       
1                                                       0       
2                                                       0       
3                                                       0       
4                                                       0       
...                                                   ...       
162978                                                  0       
162979                                                  0       
162980                                                  0       
162981                                                  0       
162982                                                  0       

       'Dangerous Permission: CLEAR_APP_USER_DATA'  \
0                                                0   
1                                                0   
2                                                0   
3                                                0   
4                                                0   
...                                            ...   
162978                                           0   
162979                                           0   
162980                                           0   
162981                                           0   
162982                                           0   

       'Dangerous Permission: DELETE_CACHE_FILES'  \
0                                               0   
1                                               0   
2                                               0   
3                                               0   
4                                               0   
...                                           ...   
162978                                          0   
162979                                          0   
162980                                          0   
162981                                          0   
162982                                          0   

       'Dangerous Permission: DELETE_PACKAGES'  \
0                                            0   
1                                            0   
2          

#### Dataset resultante contém 115 features: permissões, chamadas de API e intenções de 162983 amostras (139670 benignos + 23313 maliciosos)

In [28]:
crawlResult.to_csv("AndroCrawl_CPI.csv")